docs/designs/supporting_pytorch.md

```diff
@@ -0,0 +1,258 @@
+can use small data for debugging and verification on a single machine. If it passes,
+it can be submitted to the Kubernetes cluster for distributed fault-tolerant large-scale
+training without any code modification.
+
+The specific model building method can refer to this
+[mnist_subclass.py](https://github.com/sql-machine-learning/elasticdl/blob/develop/model_zoo/mnist/mnist_subclass.py).
+
+
+### 2. Load Data from Task
+
+ElasticDL introduces a master process for each job. By calling the Kubernetes API,
+the master process understands the cluster situation. The data is distributed by
+the master.[dynamic_data_sharding.md](https://github.com/sql-machine-learning/elasticdl/blob/develop/docs/designs/dynamic_data_sharding.md)
+
+1. A worker get a task from the master.
+2. A worker reads real data according to the offset in the task
+`feed` customizes the conversion process of training data
+to PyTorch model input.
+
+`TODO: Make DataLoader works with task, more details will be added.`
+
+There is a tutorial about [feed](https://github.com/sql-machine-learning/elasticdl/blob/dc4b2901d651cea08cdb2825a6829c97294e4652/model_zoo/mnist/mnist_subclass.py#L64)
+in TensorFlow.
+
+### 3. Transmission of Gradient Information
+
+A task received by an ElasticDL worker usually includes multiple minibatches.
+For each task, the worker opens the corresponding file or table, and then:
+1. Get a mini-batch training data.
+2. Call the user-defined `forward` function with the local model as a parameter
+to calculate the cost. If the model is large, some parameters may come from the
+parameter server.
+3. The worker performs backward calculations to obtain the gradient.
+4. In synchronous SGD, the worker calls `AllReduce` to implement FTlib to synchronize
+gradients and update the model. In asynchronous SGD, the worker uploads gradients
+to the `parameter server` from time to time, and also obtains global model parameters
+from the parameter server from time to time.
+
+```python
+while (True):
+    task = get_task()
+    dataset = create_dataset(task)
+    for minibatch in dataset:
+        pull_parameters()
+        forward()
+        backward()
+        push_gradients()
+```
+
+#### Gradient information acquisition
+
+The advanced API in PyTorch such as `torch.optim` is not available,we had to
+update the value of each parameter by name, and manually zero the gradient of
+each parameter.
+
+`torch.no_grad()` context is necessary because we don't want to record these
+operations in the next gradient calculation.
+To go further, we can use `model.parameters()` and `model.zero_grad()` (defined
+by PyTorch for `nn.Module`) to make these steps more concise, and there will be
+no errors of forgetting some parameters, especially when we build a complex model:
+
+```python
+with torch.no_grad():
+    for param in model.parameters(): 
+        param -= param.grad * lr
+    model.zero_grad()
+```
+
+#### Work with Parameter Server Client
```

Comment:
The title could be: Aggregating Gradients under Parameter Server Strategy

In the parameter server strategy,  the workers pull the latest parameters from the PS before forwarding and push gradients to the PS after backward. Then, the PS applies gradients to parameters. 

`PSClient` provides several util functions, like `push_gradients` and `pull_dense_parameters`, we could directly use them.

```
with torch.no_grad():
    grads = [param.grad.numpy() for param in model.parameters()]
    self.ps_client.push_gradients(grads)
```

Reply:
Your suggestion is very helpful.  I'm confusing the role of ps_client, which provides tools to push gradients instead of playing the role of PS.